File: backend/src/main/java/com/example/demo/models/Film.java
package com.example.demo.models;

import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "film")
@Data
@NoArgsConstructor
public class Film {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "film_id", columnDefinition = "SMALLINT UNSIGNED")
    private Short filmId;

    @Column(name = "title", nullable = false, length = 128)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "release_year", columnDefinition = "YEAR")
    private Integer releaseYear;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "language_id", nullable = false)
    private Language language;

    @Column(name = "original_language_id", columnDefinition = "TINYINT UNSIGNED")
    private Short originalLanguageId;

    @Column(name = "rental_duration", nullable = false, columnDefinition = "TINYINT UNSIGNED DEFAULT 3")
    private Short rentalDuration = 3;

    @Column(name = "rental_rate", nullable = false, precision = 4, scale = 2)
    private BigDecimal rentalRate = new BigDecimal("4.99");

    @Column(name = "length", columnDefinition = "SMALLINT UNSIGNED")
    private Short length;

    @Column(name = "replacement_cost", nullable = false, precision = 5, scale = 2)
    private BigDecimal replacementCost = new BigDecimal("19.99");

    @Column(name = "rating", columnDefinition = "ENUM('G','PG','PG-13','R','NC-17') DEFAULT 'G'")
    private String rating = "G";

    @Column(name = "special_features", columnDefinition = "SET('Trailers','Commentaries','Deleted Scenes','Behind the Scenes')")
    private String specialFeatures;

    @Column(name = "active", nullable = false)
    private Boolean active = true;

    @Column(name = "last_update", nullable = false)
    private LocalDateTime lastUpdate = LocalDateTime.now();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "film_actor",
        joinColumns = @JoinColumn(name = "film_id"),
        inverseJoinColumns = @JoinColumn(name = "actor_id")
    )
    private Set<Actor> actors = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
        name = "film_category",
        joinColumns = @JoinColumn(name = "film_id"),
        inverseJoinColumns = @JoinColumn(name = "category_id")
    )
    private Set<Category> categories = new HashSet<>();
}

File: frontend/lib/models/film.dart
class Film {
  final int? id;
  final String title;
  final String? description;
  final int? releaseYear;
  final int? languageId;
  final int? rentalDuration;
  final double? rentalRate;
  final int? length;
  final double? replacementCost;
  final String? rating;
  final String? specialFeatures;
  final bool? active;
  final List<String>? categories;
  final List<String>? actors;

  Film({
    this.id,
    required this.title,
    this.description,
    this.releaseYear,
    this.languageId,
    this.rentalDuration,
    this.rentalRate,
    this.length,
    this.replacementCost,
    this.rating,
    this.specialFeatures,
    this.active,
    this.categories,
    this.actors,
  });

  factory Film.fromJson(Map<String, dynamic> json) {
    return Film(
      id: json['id'] ?? json['filmId'],
      title: json['title'] ?? '',
      description: json['description'],
      releaseYear: json['releaseYear'],
      languageId: json['languageId'],
      rentalDuration: json['rentalDuration'],
      rentalRate: (json['rentalRate'] as num?)?.toDouble(),
      length: json['length'],
      replacementCost: (json['replacementCost'] as num?)?.toDouble(),
      rating: json['rating'],
      specialFeatures: json['specialFeatures'],
      active: json['active'] ?? json['available'],
      categories: json['categories'] != null ? List<String>.from(json['categories']) : null,
      actors: json['actors'] != null ? List<String>.from(json['actors']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'title': title,
      'description': description,
      'releaseYear': releaseYear,
      'languageId': languageId,
      'rentalDuration': rentalDuration,
      'rentalRate': rentalRate,
      'length': length,
      'replacementCost': replacementCost,
      'rating': rating,
      'specialFeatures': specialFeatures,
      'active': active,
      'categories': categories,
      'actors': actors,
    };
  }
}
